File: src/history.cc
/** 
 * @file history.cc
 * @copyright (C) 2001-2006 Marcus Bjurman\n
 * @copyright (C) 2007-2012 Piotr Eljasiak\n
 * @copyright (C) 2013-2020 Uwe Scholz\n
 *
 * @copyright This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * @copyright This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * @copyright You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
 */

#include <config.h>

#include "gnome-cmd-includes.h"
#include "history.h"
#include "utils.h"

using namespace std;


History::~History()
{
    if (ents)
    {
        g_list_foreach (ents, (GFunc) g_free, NULL);
        g_list_free (ents);
    }
}


void History::add(const gchar *text)
{
    if (is_locked)
        return;

    if (!text || !*text)
        return;

    ents = string_history_add (ents, text, max);
    pos = ents;
}


const gchar *History::first()
{
    g_return_val_if_fail (pos != NULL, NULL);

    if (pos->next)
        pos = g_list_last (pos);

    return (const gchar *) pos->data;
}


const gchar *History::back()
{
    g_return_val_if_fail (pos != NULL, NULL);

    if (pos->next)
        pos = pos->next;

    return (const gchar *) pos->data;
}


const gchar *History::forward()
{
    g_return_val_if_fail (pos != NULL, NULL);

    if (pos->prev)
        pos = pos->prev;

    return (const gchar *) pos->data;
}


const gchar *History::last()
{
    g_return_val_if_fail (pos != NULL, NULL);

    if (pos->prev)
        pos = g_list_first (pos);

    return (const gchar *) pos->data;
}

File: src/connection/history.rs
/*
 * Copyright 2025 Andrey Kutejko <andy128k@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 *
 * For more details see the file COPYING.
 */

use std::cell::{Cell, RefCell};

pub struct History<T> {
    max: usize,
    is_locked: Cell<bool>,
    entries: RefCell<Vec<T>>,
    position: Cell<usize>,
}

impl<T: PartialEq + Clone> History<T> {
    pub fn new(max: usize) -> Self {
        Self {
            max,
            is_locked: Default::default(),
            entries: Default::default(),
            position: Default::default(),
        }
    }

    pub fn add(&self, item: T) {
        if self.is_locked() {
            return;
        }

        let mut entries = self.entries.borrow_mut();
        if let Some(position) = entries.iter().position(|i| *i == item) {
            // if the same value has been given before move it first in the list
            let item = entries.remove(position);
            entries.insert(0, item);
        } else {
            // or if its new just add it
            entries.insert(0, item);
        }
        // don't let the history get too long
        entries.truncate(self.max);
        self.position.set(0);
    }

    pub fn is_locked(&self) -> bool {
        self.is_locked.get()
    }

    pub fn lock(&self) {
        self.is_locked.set(true);
    }

    pub fn unlock(&self) {
        self.is_locked.set(false);
    }

    fn current(&self) -> Option<T> {
        self.entries.borrow().get(self.position.get()).cloned()
    }

    pub fn can_back(&self) -> bool {
        self.position.get() + 1 < self.entries.borrow().len()
    }

    pub fn can_forward(&self) -> bool {
        self.position.get() > 0
    }

    pub fn first(&self) -> Option<T> {
        let len = self.entries.borrow().len();
        if len == 0 {
            return None;
        }
        self.position.set(len - 1);
        self.current()
    }

    pub fn back(&self) -> Option<T> {
        let len = self.entries.borrow().len();
        if len == 0 {
            return None;
        }

        let position = self.position.get();
        if position + 1 < len {
            self.position.set(position + 1);
            self.current()
        } else {
            None
        }
    }

    pub fn forward(&self) -> Option<T> {
        if self.entries.borrow().is_empty() {
            return None;
        }

        let position = self.position.get();
        if position > 0 {
            self.position.set(position - 1);
            self.current()
        } else {
            None
        }
    }

    pub fn last(&self) -> Option<T> {
        if self.entries.borrow().is_empty() {
            return None;
        }
        self.position.set(0);
        self.current()
    }

    pub fn export(&self) -> Vec<T> {
        self.entries.borrow().clone()
    }
}
